satpy/readers/utils.py

```diff
@@ -474,3 +474,179 @@ def remove_earthsun_distance_correction(reflectance, utc_date=None):
+            if coefs:
+                res[channel] = coefs
+        return res
+
+    def _get_coefs(self, mode_or_coefs, channel):
+        if self._is_mode(mode_or_coefs):
+            return self._get_coefs_by_mode(mode_or_coefs, channel)
+        return mode_or_coefs
+
+    def _is_mode(self, mode_or_coefs):
+        return isinstance(mode_or_coefs, str)
+
+    def _get_coefs_by_mode(self, mode, channel):
+        coefs_set = self._get_coefs_set(mode)
+        return coefs_set.get(channel, None)
+
+    def _get_coefs_set(self, mode):
+        try:
+            return self.coefs[mode]
+        except KeyError:
+            modes = list(self.coefs.keys())
+            raise KeyError(f"Unknown calibration mode: {mode}. Choose one of {modes}")
+
+    def get_calib_mode(self, wishlist, channel):
+        """Get desired calibration mode for the given channel."""
+        if isinstance(wishlist, str):
+            return wishlist
+        elif isinstance(wishlist, dict):
+            flat = self._flatten_multi_channel_keys(wishlist)
+            return flat[channel]
+
+
+class CalibrationCoefficientSelector:
+    """Helper for choosing coefficients out of multiple options.
+
+    Example: Three sets of coefficients are available (nominal, meirink, gsics).
+    A user wants to calibrate
+      - channel 1 with "meirink"
+      - channels 2/3 with "gsics"
+      - channel 4 with custom coefficients
+      - remaining channels with nominal coefficients
+
+    1. Define coefficients
+
+    .. code-block:: python
+
+        from satpy.readers.utils import CalibrationCoefficientSelector
+
+        coefs = {
+            "nominal": {
+                "ch1": "nominal_ch1",
+                "ch2": "nominal_ch2",
+                "ch3": "nominal_ch3"
+                "ch4": "nominal_ch4"
+                "ch5": "nominal_ch5"
+            },
+            "meirink": {
+                "ch1": "meirink_ch1",
+            },
+            "gsics": {
+                "ch2": "gsics_ch2",
+                # ch3 coefficients are missing
+            }
+        }
+        wishlist = {
+            "ch1": "meirink",
+            ("ch2", "ch3"): "gsics"
+            "ch4": {"mygain": 123},
+        }
+
+    2. Query:
+
+    .. code-block:: python
+
+        >>> s = CalibrationCoefficientSelector(coefs, wishlist)
+        >>> s.get_coefs("ch1")
+        "meirink_ch1"
+        >>> s.get_coefs("ch2")
+        "gsics_ch2"
+        >>> s.get_coefs("ch3")
+        KeyError: 'No gsics calibration coefficients for ch3'
+        >>> s.get_coefs("ch4")
+        {"mygain": 123}
+        >>> s.get_coefs("ch5")
+        "nominal_ch5
+
+    3. Fallback to nominal for ch3:
+
+    .. code-block:: python
+
+        >>> s = CalibrationCoefficientSelector(coefs, wishlist, fallback="nominal")
+        >>> s.get_coefs("ch3")
+        "nominal_ch3"
+    """
+
+    def __init__(self, coefs, wishlist, default="nominal", fallback=None):
```

Comment:
we have an (internal) wishlist for composites in other parts of satpy, maybe this should be called `calibration_wishlist` instead?

Reply:
Good point, done